vfio_user/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+struct DeviceGetInfo {
+    header: Header,
+    argsz: u32,
+    flags: u32,
+    num_regions: u32,
+    num_irqs: u32,
+}
+
+unsafe impl ByteValued for DeviceGetInfo {}
+
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug)]
+struct DeviceGetRegionInfo {
+    header: Header,
+    region_info: vfio_region_info,
+}
+
+unsafe impl ByteValued for DeviceGetRegionInfo {}
+
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug)]
+struct RegionAccess {
+    header: Header,
+    offset: u64,
+    region: u32,
+    count: u32,
+}
+
+unsafe impl ByteValued for RegionAccess {}
+
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug)]
+struct GetIrqInfo {
+    header: Header,
+    argsz: u32,
+    flags: u32,
+    index: u32,
+    count: u32,
+}
+
+unsafe impl ByteValued for GetIrqInfo {}
+
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug)]
+struct SetIrqs {
+    header: Header,
+    argsz: u32,
+    flags: u32,
+    index: u32,
+    start: u32,
+    count: u32,
+}
+
+unsafe impl ByteValued for SetIrqs {}
+
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug)]
+struct DeviceReset {
+    header: Header,
+}
+
+unsafe impl ByteValued for DeviceReset {}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Capabilities {
+    #[serde(default = "default_max_msg_fds")]
+    max_msg_fds: u32,
+    #[serde(default = "default_max_data_xfer_size")]
+    max_data_xfer_size: u32,
+    #[serde(default = "default_migration_capabilities")]
+    migration: MigrationCapabilities,
+}
+
+impl Default for Capabilities {
+    fn default() -> Self {
+        Self {
+            max_msg_fds: default_max_msg_fds(),
+            max_data_xfer_size: default_max_data_xfer_size(),
+            migration: default_migration_capabilities(),
+        }
+    }
+}
+
+pub struct Client {
+    stream: UnixStream,
+    next_message_id: Wrapping<u16>,
+    num_irqs: u32,
+    resettable: bool,
+    regions: Vec<Region>,
+}
+
+#[derive(Debug)]
+pub struct Region {
+    pub flags: u32,
+    pub index: u32,
+    pub size: u64,
+    pub file_offset: Option<FileOffset>,
+}
+
+#[derive(Debug)]
+pub struct IrqInfo {
+    pub index: u32,
+    pub flags: u32,
+    pub count: u32,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error connecting: {0}")]
+    Connect(#[source] std::io::Error),
+    #[error("Error serializing capabilities: {0}")]
+    SerializeCapabilites(#[source] serde_json::Error),
+    #[error("Error deserializing capabilities: {0}")]
+    DeserializeCapabilites(#[source] serde_json::Error),
+    #[error("Error writing to stream: {0}")]
+    StreamWrite(#[source] std::io::Error),
+    #[error("Error reading from stream: {0}")]
+    StreamRead(#[source] std::io::Error),
+    #[error("Error writing with file descriptors: {0}")]
+    SendWithFd(#[source] vmm_sys_util::errno::Error),
+    #[error("Error reading with file descriptors: {0}")]
+    ReceiveWithFd(#[source] vmm_sys_util::errno::Error),
+    #[error("Not a PCI device")]
+    NotPciDevice,
+}
+
+impl Client {
```

Comment:
I'm not a huge fan of using `Client` as the structure name. The reason is, the spec might let the VMM be the server in the future. So instead of client/server, maybe we should use device/backend nomenclature.
WDYT?

Reply:
Yeh I agree with the sentiment. Lets leave it for now since it matches the spec but it would be easy to change in the future.